keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/memory/InMemoryResourceRepository.kt

```diff
@@ -97,22 +100,46 @@ class InMemoryResourceRepository(
       .singleOrNull()
       ?.also {
         resources.remove(it)
-        events.remove(it)
+        resourceEvents.remove(it)
       }
       ?: throw NoSuchResourceId(id)
   }
 
+  override fun applicationEventHistory(application: String, limit: Int): List<ApplicationEvent> {
+    require(limit > 0) { "limit must be a positive integer" }
+    return applicationEvents[application]
+      ?.take(limit)
+      ?: emptyList()
+  }
+
+  override fun applicationEventHistory(application: String, downTo: Instant): List<ApplicationEvent> {
+    return applicationEvents[application]
+      ?.takeWhile { !it.timestamp.isBefore(downTo) }
+      ?: emptyList()
+  }
+
   override fun eventHistory(id: String, limit: Int): List<ResourceEvent> {
     require(limit > 0) { "limit must be a positive integer" }
-    return events[id]?.take(limit) ?: throw NoSuchResourceId(id)
+    return resourceEvents[id]
+      ?.take(limit)
+      ?: throw NoSuchResourceId(id)
   }
 
   override fun appendHistory(event: ResourceEvent) {
-    events.computeIfAbsent(event.id) {
+    appendHistory(resourceEvents, event)
+  }
+
+  override fun appendHistory(event: ApplicationEvent) {
+    appendHistory(applicationEvents, event)
+  }
+
+  private fun <T : PersistentEvent> appendHistory(eventList: MutableMap<String, MutableList<T>>, event: T) {
+    eventList.computeIfAbsent(event.uid) {
       mutableListOf()
     }
       .let {
-        if (!event.ignoreRepeatedInHistory || event.javaClass != it.firstOrNull()?.javaClass) {
+        val lastEvent = it.firstOrNull()
```

Comment:
I found this a little confusing because the variable is named `lastEvent`, but the method called is `firstOrNull`. Does "last" here mean "most recent?"

Reply:
Good point. It means "most recent". The reason I call `firstOrNull` is that the lists are always ordered in descending order.